providers/implementations/encode_decode/encode_key2any.c

```diff
@@ -841,136 +990,400 @@ static int key2any_encode(struct key2any_ctx_st *ctx, OSSL_CORE_BIO *cout,
+#define DO_PKCS8(impl, type, output)                                        \
+    DO_PRIVATE_KEY(impl, type, pkcs8, output)
 
-        if (out != NULL && writer != NULL)
-            ret = writer(out, key);
+/* SubjectPublicKeyInfo is a structure for public keys only */
+#define SubjectPublicKeyInfo_output_structure "SubjectPublicKeyInfo"
+#define DO_SubjectPublicKeyInfo_selection_mask DO_PUBLIC_KEY_selection_mask
+#define DO_SubjectPublicKeyInfo(impl, type, output)                         \
+    DO_PUBLIC_KEY(impl, type, spki, output)
 
-        BIO_free(out);
-    } else {
-        ERR_raise(ERR_LIB_PROV, ERR_R_PASSED_INVALID_ARGUMENT);
-    }
+/*
+ * "type-specific" is a uniform name for key type specific output for private
+ * and public keys as well as key parameters.  This is used internally in
+ * libcrypto so it doesn't have to have special knowledge about select key
+ * types, but also when no better name has been found.  If there are more
+ * expressive DO_ names above, those are preferred.
+ *
+ * Three forms exist:
+ *
+ * - type_specific_keypair              Only supports private and public key
+ * - type_specific_pub                  Only supports public key
+ * - type_specific_params               Only supports parameters
+ * - type_specific                      Supports all parts of an EVP_PKEY
+ * - type_specific_no_pub               Supports all parts of an EVP_PKEY
+ *                                      except public key
+ */
+#define type_specific_pub_output_structure "type-specific"
+#define DO_type_specific_pub_selection_mask DO_PUBLIC_KEY_selection_mask
+#define DO_type_specific_pub(impl, type, output)                            \
+    DO_PUBLIC_KEY(impl, type, type_specific, output)
+#define type_specific_params_output_structure "type-specific"
+#define DO_type_specific_params_selection_mask DO_PARAMETERS_selection_mask
+#define DO_type_specific_params(impl, type, output)                         \
+    DO_PARAMETERS(impl, type, type_specific, output)
+#define type_specific_keypair_output_structure "type-specific"
+#define DO_type_specific_keypair_selection_mask                             \
+    ( DO_PRIVATE_KEY_selection_mask | DO_PUBLIC_KEY_selection_mask )
+#define DO_type_specific_keypair(impl, type, output)                        \
+    DO_PRIVATE_KEY(impl, type, type_specific, output)                       \
+    DO_PUBLIC_KEY(impl, type, type_specific, output)
+#define type_specific_output_structure "type-specific"
+#define DO_type_specific_selection_mask                                     \
+    ( DO_type_specific_keypair_selection_mask                               \
+      | DO_type_specific_params_selection_mask )
+#define DO_type_specific(impl, type, output)                                \
+    DO_type_specific_keypair(impl, type, output)                            \
+    DO_type_specific_params(impl, type, output)
+#define type_specific_no_pub_output_structure "type-specific"
```

Comment:
Does this need to be an encoder type?
Wouldnt this just be part of the actual encode for EC..

If we do need it then
Are we not supporting the no_params case also?

i.e: inside i2d_ECPrivateKey()

    if (!(a->enc_flag & EC_PKEY_NO_PARAMETERS)) {
        if ((priv_key->parameters =
             EC_GROUP_get_ecpkparameters(a->group,
                                        priv_key->parameters)) == NULL) {
            ECerr(EC_F_I2D_ECPRIVATEKEY, ERR_R_EC_LIB);
            goto err;
        }
    }

Reply:
The reason is that there's no straight forward support for a type specific public key output structure in any format.  Not DER, not PEM (there is no i2d_ECPublicKey and no PEM_write_bio_ECPublicKey).

Look at it in PEM terms, most key types that have type specific format support at all usually have i2d_{TYPE}PrivakeKey, i2d_{TYPE}PublicKey, and some of them also have i2d_{TYPE}params.  That translates into `type_specific_keypair` (when there is no i2d_{TYPE}params) or `type_specific` (where there is a i2d_{TYPE}params).
The odd ones are DH (which doesn't have any `i2d_DHPrivateKey` or `i2d_DHPublicKey` and is therefore "tagged" with `type_specific_params`), and EC (which doesn't have `i2d_ECPublicKey`, and is therefore "tagged" with `type_specific_params`).

Reply:
However, this had me looking a bit closer at the different `type_specific`s, and I noticed that I can clean that up a little bit.  `type_specific_pub` isn't used anywhere.

Reply:
Cleanup done